crates/bevy_ecs/src/component.rs

```diff
@@ -201,11 +203,23 @@ pub enum StorageType {
     SparseSet,
 }
 
+/// The type used for [`Component`] lifecycle hooks such as `on_add`, `on_insert` or `on_remove`
+pub type ComponentHook = for<'w> fn(DeferredWorld<'w>, Entity, ComponentId);
+
+/// Lifecycle hooks for a given [`Component`], stored in it's [`ComponentInfo`]
+#[derive(Debug, Clone, Default)]
+pub struct ComponentHooks {
+    pub(crate) on_add: Option<ComponentHook>,
```

Comment:
It would be _really_ helpful for `bevy_ggrs` if this could be a `Vec<ComponentHook>` instead. It would allow a plugin to add hooks to a 3rd party component without fear of interrupting any possible existing behaviour. Plus, it removes the panic in registration. Since `ComponentInfo` is already 80+ bytes, the extra few for some `length` and `capacity` values in a vector aren't going to make a massive difference.

Reply:
It would also open up the possibility of replacing/removing `RemovedComponents` and `Changed` as you already mentioned.

Reply:
The size difference isn't the concern here so much as the extra indirection, hooks are performant but limited. The use case you describe is very compelling but it is better served by observers which I am putting together a PR as I type this. It should allow for an API like this:
```rust
world.observer<(&A, &B)>()
    .event<OnRemove>()
    .run(|observer| {
        for (a, b) in observer {
            // Do stuff with component data
        }
    });
```
The reason I didn't include it in this PR is this is mainly focused on setting up the concept of `DeferredWorld` and component hooks will be used in the internal implementation of observers. Check back in a couple days and I'll have something that fits the use case you want here.

Reply:
I look forward to it! If the plan is to provide an alternate API that allows multiple hooks (through Observation), then that's fine by me.

Reply:
Yep, the intention is that most users won't use hooks for things like game logic. Instead they aim to provide a way for you to setup and clean-up components that need it. Observers will be more ergonomic and flexible but in order to setup and clean-up the indexes for observers you need something like hooks so may as well expose that to users as well.

Reply:
API has evolved a little but a working example can be seen here:
https://github.com/james-j-obrien/bevy/blob/observers/examples/ecs/observers.rs